src/Stratis.SmartContracts.Core.Validation/TypePolicyValidator.cs

```diff
@@ -0,0 +1,184 @@
+    /// </summary>
+    public class ModulePolicyValidator : IModuleDefinitionValidator
+    {
+        private readonly ValidationPolicy policy;
+
+        private readonly TypePolicyValidator typePolicyValidator;
+
+        public ModulePolicyValidator(ValidationPolicy policy)
+        {
+            this.policy = policy;
+            this.typePolicyValidator = new TypePolicyValidator(policy);
+        }
+
+        public IEnumerable<ValidationResult> Validate(ModuleDefinition module)
+        {
+            var results = new List<ValidationResult>();
+
+            this.ValidateModule(results, module);
+
+            IEnumerable<TypeDefinition> contractTypes = module.GetDevelopedTypes();
+
+            foreach (TypeDefinition contractType in contractTypes)
+            {
+                results.AddRange(this.typePolicyValidator.Validate(contractType));
+            }
+            
+            return results;
+        }
+
+        private void ValidateModule(List<ValidationResult> results, ModuleDefinition module)
+        {
+            foreach (var validator in this.policy.ModuleDefValidators)
+            {
+                results.AddRange(validator.Validate(module));
+            }
+        }
+
+    }
+
+    /// <summary>
+    /// Validator for a <see cref="TypeDefinition"/> using a given <see cref="ValidationPolicy"/>
+    /// </summary>
+    public class TypePolicyValidator : ITypeDefinitionValidator
+    {
+        private readonly ValidationPolicy policy;
+
+        public TypePolicyValidator(ValidationPolicy policy)
+        {
+            this.policy = policy;
+        }
+
+        public IEnumerable<ValidationResult> Validate(TypeDefinition type)
+        {
+            var results = new List<ValidationResult>();
+
+            this.ValidateTypeDef(results, type);
+            this.ValidateNestedTypes(results, type);
+            this.ValidateFields(results, type);
+            this.ValidateMethods(results, type);
+
+            return results;
+        }
+
+        private void ValidateMethods(List<ValidationResult> results, TypeDefinition type)
+        {
+            // Validate methods
+            foreach (MethodDefinition method in type.Methods)
+            {
+                if (!method.HasBody)
+                    return;
+
+                if (method.Body.Instructions.Count == 0)
+                    return;
+
+                this.ValidateParameters(results, type, method);
+
+                foreach (var validator in this.policy.MethodDefValidators)
+                {
+                    results.AddRange(validator.Validate(method));
+                }
+
+                this.ValidateInstructions(results, type, method);               
+            }
+        }
+
+        private void ValidateParameters(List<ValidationResult> results, TypeDefinition type, MethodDefinition method)
+        {
+            if (this.policy.ParameterValidators.Any())
+            {
```

Comment:
Maybe just return here ` if (!this.policy.ParameterValidators.Any())` ? Then we eliminate some of these nested ifs.